feat(runtime): support declarative shadow DOM

christian-bromann opened the pull request:
## What is the current behavior?
This patch introduces support for declarative shadow DOM in Stencil 🎉 

GitHub Issue Number: #4010

## What is the new behavior?
This patch includes:

- __New Feature:__ new new flags to the `renderToString` method:
  - `fullDocument`: allows to define whether to return a full document or just the component
  - `serializeShadowRoot`: defines whether a component marked with `shadow: true` is being rendered in a declarative shadow DOM:
      ```ts
      <my-cmp>
        <template shadowrootmode="open">
          <style>...</style>
          <div>Hello World</div>
          <slot></slot>
        </template>

        Slotted content
      </my-cmp>
      ```
    or as scoped component:
    ```ts
    <my-cmp>
      <div>Hello World</div>
      Slotted content
    </my-cmp>
    ```
- __New Feature:__ the hydrate module now exports a `streamToString` function that returns a `Readable` object that can be passed into a server response.
- __Potentially Breaking:__ users doing DOM snapshot tests of shadow components will need to update their snapshots as all shadow components are now serialized in a Declarative Shadow DOM
- Runtime changes as suggested in #5787
- ESM version of the hydrate script
- Removal of `isPromise` helper in the code
- Support for objects in component properties, e.g. `renderToString(`<car-detail car=${JSON.stringify({ year: 1234 })}></car-detail>`)`
- Removal of duplicated logic within `renderToString` and `hyrdrateDocument`
- Rename of `serializeToHtml` to `streamToHtml` which uses generators
- some changes to how we call `setTimeout` or `setInterval` in mock doc in order to enable support for using the hydrate script in the browser

## Documentation

__ToDo__ missing

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Testing

Added e2e tests for using the hydration script in the browser as well as part of the e2e test suite.

## Other information

n/a

tanner-reits reviewed (changes requested):
Ran into one issue when testing this out. I setup a test app that used the hydrate output from a Stencil starter project. Only change I made to the starter was adding a `slot` element in the `render` function return value for `my-component`. Then, if I added the following snippet to my test server, it would result in 3 instances of the component getting rendered:

```ts
const res = await renderToString(
    '<my-component>Jimmy</my-component><my-component first="Bob"></my-component>'
  );
  response.writeHead(200);
  response.write(res.html);
  response.end();
```

The first two elements that get rendered are correct, but the third is an instance without any slot content or input property values. Same thing happens when using `streamToString`. I checked with a published version of Stencil and didn't see this behavior, so something we might wanna look into. I'm sure it's an issue client-side when the runtime takes over since the server wasn't spitting out any additional elements in the strings it returned.

Also, code looks good, but would be good to cut down the number of new SNC violations this introduces :)

christian-bromann commented on test/end-to-end/src/declarative-shadow-dom/test.e2e.ts:
```diff
@@ -0,0 +1,196 @@
+  });
+}
+
+jest.setTimeout(3000000);
```
ups 🙈 left this in for debugging

christian-bromann commented:
@tanner-reits thanks for reviewing!

I am struggling reproducing the issue you are describing. I've set-up a new Stencil project with the following component:

```tsx
import { Component, Prop, h } from '@stencil/core';
import { format } from '../../utils/utils';

@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  shadow: true,
})
export class MyComponent {
  /**
   * The first name
   */
  @Prop() first: string;

  /**
   * The middle name
   */
  @Prop() middle: string;

  /**
   * The last name
   */
  @Prop() last: string;

  private getText(): string {
    return format(this.first, this.middle, this.last);
  }

  render() {
    return <div>
      Hello, World! I'm {this.getText()}
      <slot></slot>
    </div>;
  }
}
```
When creating an hydration script and calling this script:

```js
const { renderToString } = require('./hydrate');

(async () => {
  console.log(await renderToString('<my-component>Jimmy</my-component><my-component first="Bob"></my-component>', {
    prettyHtml: true,,
    serializeShadowRoot: true
  }));
})()
```

I correctly get the following output:

```html
<!doctype html>
<html class="hydrated" data-stencil-build="u56d8u2z">
  <head>
    <meta charset="utf-8">
  </head>
  <body>
    <my-component class="hydrated sc-my-component-h" s-id="1">
      <template shadowrootmode="open">
        <style sty-id="sc-my-component">
          /*!@:host*/.sc-my-component-h{display:block}
        </style>
        <div c-id="1.0.0.0" class="sc-my-component">
          <!--t.1.1.1.0-->
          Hello, World! I'm
          <slot c-id="1.2.1.1" class="sc-my-component"></slot>
        </div>
      </template>
      <!--r.1-->
      Jimmy
    </my-component>
    <my-component class="hydrated sc-my-component-h" first="Bob" s-id="2">
      <template shadowrootmode="open">
        <style sty-id="sc-my-component">
          /*!@:host*/.sc-my-component-h{display:block}
        </style>
        <div c-id="2.0.0.0" class="sc-my-component">
          <!--t.2.1.1.0-->
          Hello, World! I'm Bob
          <slot c-id="2.2.1.1" class="sc-my-component"></slot>
        </div>
      </template>
      <!--r.2-->
    </my-component>
  </body>
</html>
```

Which just shows two elements. Mind sharing your example?

tanner-reits commented:
@christian-bromann The `renderToString` method is generating the correct output, but it doesn't render correctly when handed over to the browser

tanner-reits reviewed (approved):
LGTM! `renderToString` and `streamToString` are both emitting the expected HTML and the browser is rendering correctly!

christian-bromann commented:
@mayerraphael thanks so much for providing feedback 🙏  is there any chance you can create a minimal reproducible example? I will try to reproduce this myself but haven't come across this behavior.

christian-bromann commented:
@mayerraphael thanks for reporting, I proposed a fix and made a dev release (`4.18.3-dev.1719344120.e32bcd8`) which resolves the issue.

thure commented:
@christian-bromann I was working from babe807b5b72a46bfd9bd8e9c9b4cda962e25607, which correctly hydrated my components, however none of the releases on NPM will hydrate the components, they instead seem not to recognize the components.

In my use-case:
```ts
  const { html } = await renderToString(
  '<ch-oklch-picker lightness="0.43" chroma="0.4" hue="256"></ch-oklch-picker>'
  {
    serializeShadowRoot: true,
    fullDocument: false,
  });
```

should return what babe807 returned:

```html
<ch-oklch-picker lightness="0.43" chroma="0.4" hue="256"></ch-oklch-picker> <ch-oklch-picker lightness="0.43" chroma="0.4" hue="256" role="group" class="hydrated" s-id="1"><!--r.1--><label id="hue-0329" c-id="1.0.0.0"><!--t.1.1.1.0-->Hue (0–360)</label><input property="hue" aria-labelledby="hue-0329" type="number" min="0" max="360" step="1" value="256" c-id="1.2.0.1"><input property="hue" aria-labelledby="hue-0329" type="range" min="0" max="360" step="1" value="256" c-id="1.3.0.2"><label id="chroma-66e5" c-id="1.4.0.3"><!--t.1.5.1.0-->Chroma (0–0.4)</label><input property="chroma" aria-labelledby="chroma-66e5" type="number" min="0.000" max="0.400" step="0.004" value="0.4" c-id="1.6.0.4"><input property="chroma" aria-labelledby="chroma-66e5" type="range" min="0.000" max="0.400" step="0.004" value="0.4" c-id="1.7.0.5"><label id="lightness-4fd3" c-id="1.8.0.6"><!--t.1.9.1.0-->Lightness (0–1)</label><input property="lightness" aria-labelledby="lightness-4fd3" type="number" min="0.00" max="1.00" step="0.01" value="0.43" c-id="1.10.0.7"><input property="lightness" aria-labelledby="lightness-4fd3" type="range" min="0.00" max="1.00" step="0.01" value="0.43" c-id="1.11.0.8"></ch-oklch-picker>
```

However 4.19.0 and any of the prereleases just return the input string.

Did the API change, or is there something I should be doing to configure the hydrate app properly?

thure commented:
I’ve posted a comparison PR for the hydrate outputs of babe807 and 4.19.0 here:
https://github.com/thure/hydrate-diff/pull/1/files?diff=split&w=1

Could one of these differences account for the component not hydrating?

christian-bromann commented:
> Did the API change, or is there something I should be doing to configure the hydrate app properly?

No, we've build it to be backward compatible. Let me take a look.

christian-bromann commented:
@thure it seems like setting up your project and the branch and running this script:
```js
import { renderToString } from './packages/elements-hydrate-temp/index.mjs'

const { html } = await renderToString(
  '<ch-oklch-picker lightness="0.43" chroma="0.4" hue="256"></ch-oklch-picker>',
  {
    serializeShadowRoot: true,
    fullDocument: false,
  }
);

console.log(html);
```
gives me above mentioned hydrated string. Can you provide some concrete steps I can walk through to properly reproduce what you see?

christian-bromann commented:
> I updated my repository again. I couldn't replicate the missing s-ids. But i got another bug with s-ids beeing set on child-components.

@mayerraphael thanks again for your feedback, I check out your repository, updated Stencil to `v4.19.0` which we released yesterday and ran `server.js` which returned the following HTML code:

```html
<my-component last-page="5" class="sc-my-component-h hydrated" s-id="1">
    <template shadowrootmode="open">
      <style sty-id="sc-my-component">
        /*!@:host*/
        .sc-my-component-h {
          display: block
        }
      </style>
      <div class="sc-my-component" c-id="1.0.0.0">
        <div class="pagination sc-my-component" c-id="1.1.1.0">
          <div class="pagination-pages pagination-notation sc-my-component" c-id="1.2.2.0">
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="1.3.3.0" s-id="2">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="2.0.0.0">
                  <!--t.2.1.1.0-->0
                </div>
              </template>
              <!--r.2-->
            </my-other-component>
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="1.4.3.1" s-id="3">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="3.0.0.0">
                  <!--t.3.1.1.0-->1
                </div>
              </template>
              <!--r.3-->
            </my-other-component>
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="1.5.3.2" s-id="4">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="4.0.0.0">
                  <!--t.4.1.1.0-->2
                </div>
              </template>
              <!--r.4-->
            </my-other-component>
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="1.6.3.3" s-id="5">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="5.0.0.0">
                  <!--t.5.1.1.0-->3
                </div>
              </template>
              <!--r.5-->
            </my-other-component>
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="1.7.3.4" s-id="6">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="6.0.0.0">
                  <!--t.6.1.1.0-->4
                </div>
              </template>
              <!--r.6-->
            </my-other-component>
          </div>
        </div>
      </div>
    </template>
    <!--r.1-->
  </my-component>
  <div>
    <my-other-component label="2" class="sc-my-other-component-h hydrated" s-id="7">
      <template shadowrootmode="open">
        <style sty-id="sc-my-other-component">
          /*!@:host*/
          .sc-my-other-component-h {
            display: block
          }
        </style>
        <div class="pagination-item sc-my-other-component" c-id="7.0.0.0">
          <!--t.7.1.1.0-->2
        </div>
      </template>
      <!--r.7-->
    </my-other-component>
  </div>
  <my-component last-page="2" class="sc-my-component-h hydrated" s-id="8">
    <template shadowrootmode="open">
      <style sty-id="sc-my-component">
        /*!@:host*/
        .sc-my-component-h {
          display: block
        }
      </style>
      <div class="sc-my-component" c-id="8.0.0.0">
        <div class="pagination sc-my-component" c-id="8.1.1.0">
          <div class="pagination-pages pagination-notation sc-my-component" c-id="8.2.2.0">
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="8.3.3.0" s-id="9">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="9.0.0.0">
                  <!--t.9.1.1.0-->0
                </div>
              </template>
              <!--r.9-->
            </my-other-component>
            <my-other-component class="sc-my-component sc-my-other-component-h hydrated" c-id="8.4.3.1" s-id="10">
              <template shadowrootmode="open">
                <style sty-id="sc-my-other-component">
                  /*!@:host*/
                  .sc-my-other-component-h {
                    display: block
                  }
                </style>
                <div class="pagination-item sc-my-other-component" c-id="10.0.0.0">
                  <!--t.10.1.1.0-->1
                </div>
              </template>
              <!--r.10-->
            </my-other-component>
          </div>
        </div>
      </div>
    </template>
    <!--r.8-->
  </my-component>
  <script type="module">
    import {
      defineCustomElements
    } from "./static/loader/index.js";
    defineCustomElements().catch(console.error);
  </script>
```

I can see all `c-id`s properly assigned. I can't find any of the patternlib components though so you might look at a different example. Can you get something reproducible for me? Thank you!

thure commented:
I should note that `ch-oklch-picker` is a `shadow: false` component — did 4.19.0 drop support for hydrating light DOM components?

christian-bromann commented:
> did 4.19.0 drop support for hydrating light DOM components?

No!

> then repeat steps 5 and 6, the build will have the hydrated component.

Unfortunately I am getting this error:
```
> nx run icons:build-esm

✘ [ERROR] No loader is configured for ".node" files: node_modules/.pnpm/@resvg+resvg-js-darwin-arm64@2.6.2/node_modules/@resvg/resvg-js-darwin-arm64/resvgjs.darwin-arm64.node

    node_modules/.pnpm/@resvg+resvg-js@2.6.2/node_modules/@resvg/resvg-js/js-binding.js:1:2588:
      1 │ ...eBinding=require("@resvg/resvg-js-darwin-arm64")}catch(e){loadEr...
        ╵                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

christian-bromann commented:
@thure I think I was able to reproduce this using internal test infrastructure.

christian-bromann commented:
@thure false alarm 🙈 it does seem with our current tests that scoped components hydrate just fine. Mind providing more minimalistic example of what you experience?

thure commented:
@christian-bromann I’ve set up [a Codesandbox with the hydrate apps](https://codesandbox.io/p/sandbox/ch-ui-debug-hydrate-app-mnjnhm?file=%2Fsrc%2Findex.ts%3A17%2C19&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxxtes7v00063b6jmfxnx1gy%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxxtes7v00023b6j8myux7qm%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxxtes7v00033b6jslh5vp5e%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxxtes7v00053b6jgyj3g0vb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxxtes7v00023b6j8myux7qm%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxxtes7u00013b6jzf26seeu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A17%252C%2522startColumn%2522%253A19%252C%2522endLineNumber%2522%253A17%252C%2522endColumn%2522%253A19%257D%255D%257D%255D%252C%2522id%2522%253A%2522clxxtes7v00023b6j8myux7qm%2522%252C%2522activeTabId%2522%253A%2522clxxtes7u00013b6jzf26seeu%2522%257D%252C%2522clxxtes7v00053b6jgyj3g0vb%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxxtes7v00043b6jiyt79cm1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxxtes7v00053b6jgyj3g0vb%2522%252C%2522activeTabId%2522%253A%2522clxxtes7v00043b6jiyt79cm1%2522%257D%252C%2522clxxtes7v00033b6jslh5vp5e%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxxtes7v00033b6jslh5vp5e%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) produced by both babe807 and 4.19.0 — these are the same files as the [diff I provided earlier](https://github.com/thure/hydrate-diff/pull/1/files?diff=split&w=1).

If you have the `ch-ui` repo locally and want to build the hydrate app yourself, you could `cd packages/elements` and run `pnpm link $pathToYourLocalStencil`, then switch to different versions in your local stencil project and run `pnpm nx build elements` in the root of `ch-ui` to see the different outputs.

christian-bromann commented:
@mayerraphael thanks a lot! I will take a look.
